File: Ising_Model/Project.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <random>
#include <numeric>
#include <fstream>
#include <filesystem>
using namespace std;

double get_energy(vector<vector<int>> lattice);
pair<vector<double>, vector<double>> metropolis(vector<vector<int>> lattice, int times, double bj, double energy);
vector<vector<double>> make_lattice(int N);
vector<vector<int>> convolve(vector<vector<int>> lattice);
double total_sum(const vector<vector<int>>& lattice);

//N x N grid
    const int N = 50;

int main(){

    const int times = 10000;
    
    auto lattice_po = make_lattice(N);
    vector<vector<int>> lattice_p(N, vector<int>(N, 0));
    for(size_t i = 0; i < lattice_po.size(); i++){
        for(size_t j = 0; j < lattice_po[i].size(); j++){
            lattice_po[i][j] > 0.75 ? lattice_p[i][j] = -1 : lattice_p[i][j] = 1;
        }
    }
    
    auto lattice_no = make_lattice(N);
    vector<vector<int>> lattice_n(N, vector<int>(N, 0));
    for(size_t i = 0; i < lattice_no.size(); i++){
        for(size_t j = 0; j < lattice_no[i].size(); j++){
            lattice_no[i][j] > 0.75 ? lattice_n[i][j] = 1 : lattice_n[i][j] = -1;
        }
    }
    
    auto result = metropolis(lattice_p, times, 0.2, get_energy(lattice_p));
    vector<double> spinsp = result.first;
    vector<double> energiesp = result.second;
    
    ofstream filep("C:\\Users\\Jason\\Desktop\\Study Material\\PH1050\\Project\\datap.txt");
    double count = 0;
    for(double i : spinsp){
        filep << i/N*N << ' ' << energiesp[count] << ' ' << count << endl;
        count++;
    }
    filep.close();
    
    result = metropolis(lattice_n, times, 0.2, get_energy(lattice_n));
    vector<double> spinsn = result.first;
    vector<double> energiesn = result.second;
    
    ofstream filen("C:\\Users\\Jason\\Desktop\\Study Material\\PH1050\\Project\\datan.txt");
    count = 0;
    for(double i : spinsn){
        filen << i/N*N << ' ' << energiesn[count] << ' ' << count << endl;
        count++;
    }
    filen.close();
    
    ofstream filebj("C:\\Users\\Jason\\Desktop\\Study Material\\PH1050\\Project\\databj.txt");
    vector<double> bjs;
    for(double bj = 0.1; bj<=2; bj+=0.05){
        auto metrop = metropolis(lattice_p, times, bj, get_energy(lattice_p));
        vector<double> spinsp = metrop.first;
        double avgSpinp = accumulate(spinsp.begin(), spinsp.end(), 0.0) / (spinsp.size()*N*N);
        auto metron = metropolis(lattice_n, times, bj, get_energy(lattice_n));
        vector<double> spinsn = metron.first;
        double avgSpinn = accumulate(spinsn.begin(), spinsn.end(), 0.0) / (spinsn.size()*N*N);
        filebj << bj << ' ' << avgSpinp << ' ' << avgSpinn << endl;
    }
    filebj.close();

}

vector<vector<double>> make_lattice(int N){
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<> dist(0.0, 1.0);
    vector<vector<double>> v(N, vector<double>(N,0));
    for(int i = 0; i < N; i++){
        for(int j = 0; j < N; j++){
            v[i][j] = dist(gen);
        }
    }
    return v;

}

double total_sum(const vector<vector<int>>& lattice){
    double sum = 0;
    for(auto row : lattice){
        sum += accumulate(row.begin(), row.end(), 0.0);
    }
    return sum;
}


double get_energy(vector<vector<int>> lattice){
    auto conv = convolve(lattice);
    double e = 0.0;
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            e += lattice[i][j] * conv[i][j];
    return -e;
}

pair<vector<double>, vector<double>> metropolis(vector<vector<int>> lattice, int times, double bj, double energy){
    vector<double> spins;
    vector<double> energies;
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dist_int(0, N-1);
    uniform_real_distribution<> prob(0.0, 1.0);
    for(int i = 0; i < times; i++){
        int x = dist_int(gen);
        int y = dist_int(gen);
        int spin_i = lattice[x][y];
        int spin_f = -spin_i;
        double E_i = 0;
        double E_f = 0;
        
        if (x>0){
            E_i += -spin_i*lattice[x-1][y];
            E_f += -spin_f*lattice[x-1][y];
        }
        if (x<N-1){
            E_i += -spin_i*lattice[x+1][y];
            E_f += -spin_f*lattice[x+1][y];
        }
        if (y>0){
            E_i += -spin_i*lattice[x][y-1];
            E_f += -spin_f*lattice[x][y-1];
        }
        if (y<N-1){
            E_i += -spin_i*lattice[x][y+1];
            E_f += -spin_f*lattice[x][y+1];
        }
        double dE = E_f - E_i;
        
        
        if(dE > 0 && (prob(gen) < exp(-bj*dE))){
            lattice[x][y] = spin_f;
            energy += dE;
        }
        else if(dE < 0){
            lattice[x][y] = spin_f;
            energy += dE;
        }

        spins.push_back(total_sum(lattice));
        energies.push_back(energy);
    }
    return make_pair(spins, energies);



}

vector<vector<int>> convolve(vector<vector<int>> lattice){
    int kern[3][3] = {{0, 1, 0},
                      {1, 0, 1},
                      {0, 1, 0}};
    vector<vector<int>> conv(N, vector<int>(N, 0));
    for(size_t i = 0; i < lattice.size(); i++){
        for(size_t j = 0; j < lattice.size(); j++){
            int sum = 0;
            for(int n = 0; n<3; n++){
                for(int m = 0; m<3; m++){
                    if(i+n-1 < 0 || j+m-1 < 0 || i+n-1 >= N || j+m-1 >= N){
                        sum += 0;
                    }
                    else{
                        sum += lattice[i+n-1][j+m-1]*kern[n][m];
                    }
            }   
            }
            conv[i][j] = sum;
    }   
    }
    return conv;
}

File: Ising_Model/Project.py
import numpy as np
import numba
import matplotlib.pyplot as plt
from numba import njit
from scipy.ndimage import convolve, generate_binary_structure

#NxN grid
N = 50

init_random = np.random.random((N,N))
lattice_n = np.zeros((N,N))
lattice_n[init_random>=0.75] = 1
lattice_n[init_random<0.75] = -1

init_random = np.random.random((N,N))
lattice_p = np.zeros((N, N))
lattice_p[init_random>=0.25] = 1
lattice_p[init_random<0.25] = -1

def get_energy(lattice):
    kern = generate_binary_structure(2,1)
    kern[1][1] = False
    arr = -lattice * convolve(lattice, kern, mode='constant', cval = 0)
    return arr.sum()


@numba.njit("UniTuple(f8[:], 2)(f8[:,:], i8, f8, f8)", nopython=True, nogil=True)
def metropolis(spin_arr, times, BJ, energy):
    spin_arr = spin_arr.copy()
    net_spins = np.zeros(times-1)
    net_energy = np.zeros(times-1)
    for t in range(0, times-1):
        x = np.random.randint(0,N)
        y = np.random.randint(0,N)
        spin_i = spin_arr[x,y] #initial spin
        spin_f = spin_i*-1 #proposed spin flip
        
        E_i = 0
        E_f = 0
        if x>0:
            E_i += -spin_i*spin_arr[x-1,y]
            E_f += -spin_f*spin_arr[x-1,y]
        if y>0:
            E_i += -spin_i*spin_arr[x,y-1]
            E_f += -spin_f*spin_arr[x,y-1]
        if x<N-1:
            E_i += -spin_i*spin_arr[x+1,y]
            E_f += -spin_f*spin_arr[x+1,y]
        if y<N-1:
            E_i += -spin_i*spin_arr[x,y+1]
            E_f += -spin_f*spin_arr[x,y+1]

        dE = E_f-E_i
        if (dE>0)*(np.random.random() < np.exp(-BJ*dE)):
            spin_arr[x,y] = spin_f
            energy += dE
        elif dE<=0:
            spin_arr[x,y] = spin_f
            energy += dE

        net_spins[t] = spin_arr.sum()
        net_energy[t] = energy

    return net_spins, net_energy

def plot(spins, energies, N):
    fig, axes = plt.subplots(1, 2, figsize=(12,4))
    ax = axes[0]
    ax.plot(spins/N**2)
    ax.set_xlabel('Algorithm Time Steps')
    ax.set_ylabel(r'Average Spin $\bar{m}$')
    ax.grid()
    ax = axes[1]
    ax.plot(energies)
    ax.set_xlabel('Algorithm Time Steps')
    ax.set_ylabel(r'Energy $E/J$')
    ax.grid()
    fig.tight_layout()
    fig.suptitle(r'Evolution of Average Spin and Energy for $\beta J=$0.7', y=1.07, size=18)
    plt.show()
spins, energies = metropolis(lattice_n, 100000, 0.2, get_energy(lattice_n))
plot(spins,energies,N)

filep = open("datap.txt", "w")
for i in range (99999):
    spinp = str(spins[i])
    energyp = str(energies[i])
    string = spinp + ' ' + energyp + ' ' + str(i) +'\n'
    filep.write(string)
filep.close()

spins, energies = metropolis(lattice_p, 100000, 0.2, get_energy(lattice_p))
plot(spins,energies,N)

filen = open("datan.txt", "w")
for i in range (99999):
    spinn = str(spins[i])
    energyn = str(energies[i])
    string = spinn + ' ' + energyn + ' ' + str(i) + '\n'
    filen.write(string)
filen.close()



def get_spin_energy(lattice, BJs):
    ms = np.zeros(len(BJs))
    E_means = np.zeros(len(BJs))
    E_stds = np.zeros(len(BJs))
    for i, bj in enumerate(BJs):
        spins, energies = metropolis(lattice, 100000, bj, get_energy(lattice))
        ms[i] = spins[-100000: ].mean()/N**2
        E_means[i] = energies[-100000:].mean()
        E_stds[i] = energies[-100000:].std()
    return E_means, E_stds, ms

BJs = np.arange(0.1, 2, 0.05)
E_means_p, E_stds_p, ms_p = get_spin_energy(lattice_p, BJs)
E_means_n, E_stds_n, ms_n = get_spin_energy(lattice_n, BJs)

plt.figure(figsize=(8,5))
plt.plot(1/BJs, ms_n, 'o--',color = 'black', label='75% of spins started negative')
plt.plot(1/BJs, ms_p, 'o--',color = 'red',  label='75% of spins started positive')
plt.xlabel(r'$\left(\frac{k}{J}\right)T$')
plt.ylabel(r'$\bar{m}$')
plt.legend(facecolor='white', framealpha=1)
plt.show()
